File: src/main/java/ufrn/imd/concorrente/TfIdfCalculator.java
package ufrn.imd.concorrente;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.HashSet;
import java.util.Set;

public class TfIdfCalculator {

    public static double calculateTf(String term, List<String> document) {
        long termCount = 0;
        for (String s : document) {
            if (s.equalsIgnoreCase(term)) {
                termCount++;
            }
        }
        return (double) termCount / document.size();
    }

    public static double calculateIdf(String term, int totalDocuments, Map<String, Integer> documentFrequencyMap) {
        int documentsContainingTerm = documentFrequencyMap.getOrDefault(term.toLowerCase(), 0);

        if (documentsContainingTerm == 0) {
            return 0.0;
        }
        return Math.log((double) totalDocuments / documentsContainingTerm);
    }

    public static double calculateTfIdf(String term, List<String> document, int totalDocuments, Map<String, Integer> documentFrequencyMap) {
        double tf = calculateTf(term, document);
        double idf = calculateIdf(term, totalDocuments, documentFrequencyMap);
        return tf * idf;
    }

    public static List<String> tokenize(String line) {
        String[] words = line.toLowerCase()
                .replaceAll("[^a-záéíóúâêîôûãõç\\s]", "")
                .trim()
                .split("\\s+");
        List<String> tokens = new ArrayList<>();
        for (String word : words) {
            if (!word.isEmpty()) {
                tokens.add(word.intern());
            }
        }
        return tokens;
    }

    public static void main(String[] args) {
        String filePath = "./dataset2.txt";
        List<List<String>> documents = new ArrayList<>();
        List<String> currentDocumentWords = new ArrayList<>();

        System.out.println("Iniciando leitura e tokenização dos documentos...");
        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            String line;
            while ((line = br.readLine()) != null) {
                if (line.trim().isEmpty()) {
                    if (!currentDocumentWords.isEmpty()) {
                        documents.add(new ArrayList<>(currentDocumentWords));
                        currentDocumentWords.clear();
                    }
                } else {
                    currentDocumentWords.addAll(tokenize(line));
                }
            }
            if (!currentDocumentWords.isEmpty()) {
                documents.add(currentDocumentWords);
            }
        } catch (IOException e) {
            System.err.println("Erro ao ler o arquivo: " + e.getMessage());
            e.printStackTrace();
            return;
        }

        if (documents.isEmpty()) {
            System.out.println("Nenhum documento encontrado no arquivo.");
            return;
        }
        System.out.println("Número de documentos lidos: " + documents.size());
        System.out.println("Pré-calculando frequências de documentos (DF)...");

        
        Map<String, Integer> documentFrequencyMap = new HashMap<>();
        Set<String> allUniqueTerms = new HashSet<>();

        for (List<String> doc : documents) {
            Set<String> termsInCurrentDoc = new HashSet<>();
            for (String token : doc) {
                allUniqueTerms.add(token);
                termsInCurrentDoc.add(token);
            }
            for (String term : termsInCurrentDoc) {
                documentFrequencyMap.put(term, documentFrequencyMap.getOrDefault(term, 0) + 1);
            }
        }
        System.out.println("Número total de termos únicos no corpus: " + allUniqueTerms.size());
        System.out.println("Cálculo de DF concluído.");

        List<Map<String, Double>> tfIdfScoresPerDocument = new ArrayList<>(); 

        System.out.println("\nCalculando TF-IDF para cada termo em cada documento...");
        for (int i = 0; i < documents.size(); i++) {
            List<String> doc = documents.get(i);
            Map<String, Double> tfIdfScores = new HashMap<>();

            Set<String> uniqueTermsInDoc = new HashSet<>(doc);

            if (i % 100 == 0 && i > 0) {
                System.out.println("Processando documento " + (i+1) + "/" + documents.size());
            }

            for (String term : uniqueTermsInDoc) {
                if (term.trim().isEmpty()) continue;
                double tfIdf = calculateTfIdf(term, doc, documents.size(), documentFrequencyMap);
                if (tfIdf > 0) {
                    tfIdfScores.put(term, tfIdf);
                }
            }

            tfIdfScoresPerDocument.add(tfIdfScores); 
        }
        System.out.println("Cálculo de TF-IDF concluído.");

        if (!tfIdfScoresPerDocument.isEmpty()) {
            System.out.println("\nExemplo de scores TF-IDF para o primeiro documento:");
            tfIdfScoresPerDocument.get(0).entrySet().stream()
                    .sorted(Map.Entry.<String, Double>comparingByValue().reversed())
                    .limit(10) 
                    .forEach(entry -> System.out.printf("Termo: '%s', TF-IDF: %.4f\n", entry.getKey(), entry.getValue()));
        }
    }
}

File: scala/TfIdfCalculator.scala
import scala.io.Source
import scala.util.Using
import scala.collection.mutable
import scala.math.log

object TfIdfCalculator {

  // A função tokenize permanece a mesma
  def tokenize(line: String): List[String] = {
    line.toLowerCase
      .replaceAll("[^a-záéíóúâêîôûãõç\\s]", "")
      .trim
      .split("\\s+")
      .filterNot(_.isEmpty)
      .toList // Removi .intern() para um teste, pois pode ser um gargalo
  }

  // As funções de cálculo de TF e IDF permanecem as mesmas
  def calculateTf(term: String, termCounts: Map[String, Int], docSize: Int): Double = {
    if (docSize == 0) 0.0
    else termCounts.getOrElse(term, 0).toDouble / docSize
  }

  def calculateIdf(term: String, totalDocuments: Int, documentFrequencyMap: Map[String, Int]): Double = {
    val documentsContainingTerm = documentFrequencyMap.getOrElse(term, 0)
    if (documentsContainingTerm == 0) 0.0
    else log(totalDocuments.toDouble / documentsContainingTerm)
  }


  def main(args: Array[String]): Unit = {
    val filePath = "src/main/java/ufrn/imd/concorrente/dataset2.txt"

    println("Iniciando a PRIMEIRA PASSAGEM para calcular DF (Frequência de Documento)...")

    // --- PRIMEIRA PASSAGEM: Calcular DF e total de documentos ---
    // Usamos Maps mutáveis para eficiência nesta passagem
    val documentFrequencyMap = mutable.Map[String, Int]().withDefaultValue(0)
    var totalDocuments = 0

    Using(Source.fromFile(filePath)) { source =>
      val lineIterator = source.getLines()
      var currentDocumentWords = mutable.ListBuffer[String]()

      for (line <- lineIterator) {
        if (line.trim.isEmpty) {
          if (currentDocumentWords.nonEmpty) {
            totalDocuments += 1
            // Pega os termos únicos e atualiza o mapa de DF
            val uniqueTermsInDoc = currentDocumentWords.toSet
            for (term <- uniqueTermsInDoc) {
              documentFrequencyMap(term) += 1
            }
            currentDocumentWords.clear() // Libera memória do documento
          }
        } else {
          currentDocumentWords ++= tokenize(line)
        }
      }
      // Processa o último documento do arquivo
      if (currentDocumentWords.nonEmpty) {
        totalDocuments += 1
        val uniqueTermsInDoc = currentDocumentWords.toSet
        for (term <- uniqueTermsInDoc) {
          documentFrequencyMap(term) += 1
        }
      }
    } match {
      case scala.util.Failure(e) =>
        System.err.println(s"Erro na primeira passagem: ${e.getMessage}")
        e.printStackTrace()
        return
      case _ =>
    }

    println(s"Primeira passagem concluída. Total de documentos: $totalDocuments. Termos únicos: ${documentFrequencyMap.size}")
    val immutableDfMap = documentFrequencyMap.toMap // Converte para imutável para a segunda passagem

    println("\nIniciando a SEGUNDA PASSAGEM para calcular TF-IDF...")

    // --- SEGUNDA PASSAGEM: Calcular TF-IDF e mostrar resultados ---
    var docsProcessed = 0
    Using(Source.fromFile(filePath)) { source =>
      val lineIterator = source.getLines()
      var currentDocumentWords = mutable.ListBuffer[String]()

      for (line <- lineIterator) {
        if (line.trim.isEmpty) {
          if (currentDocumentWords.nonEmpty) {
            // Processa o documento completo
            val docTokens = currentDocumentWords.toList
            val docSize = docTokens.length
            val termCounts = docTokens.groupBy(identity).view.mapValues(_.size).toMap

            val tfIdfScores = termCounts.keySet.map { term =>
              val tf = calculateTf(term, termCounts, docSize)
              val idf = calculateIdf(term, totalDocuments, immutableDfMap)
              (term, tf * idf)
            }.filter(_._2 > 0).toMap

            // Exemplo: Imprime os scores para o primeiro documento
            if (docsProcessed == 0) {
              println("\nExemplo de scores TF-IDF para o primeiro documento (top 10):")
              tfIdfScores.toList
                .sortBy(-_._2) // Ordena por score descendente
                .take(10)
                .foreach { case (term, score) =>
                  printf("Termo: '%s', TF-IDF: %.4f\n", term, score)
                }
            }

            docsProcessed += 1
            if (docsProcessed % 100 == 0) {
              println(s"Processando documento $docsProcessed/$totalDocuments")
            }
            currentDocumentWords.clear() // Libera a memória do documento
          }
        } else {
          currentDocumentWords ++= tokenize(line)
        }
      }

      // Processa o último documento do arquivo
      if (currentDocumentWords.nonEmpty) {
        val docTokens = currentDocumentWords.toList
        val docSize = docTokens.length
        val termCounts = docTokens.groupBy(identity).view.mapValues(_.size).toMap
        // ... (lógica de cálculo e impressão de TF-IDF, como acima)
        docsProcessed += 1
      }
    }

    println(s"\nCálculo de TF-IDF concluído. Total de documentos processados: $docsProcessed.")
  }
}